File: src/main/kotlin/mcp/MCPPythonClient.kt
package org.kozyrev.mcp

import kotlinx.coroutines.*
import kotlinx.serialization.json.*
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader

/**
 * MCP клиент, который использует Python FastMCP библиотеку для подключения
 *
 * Этот клиент запускает Python скрипт для взаимодействия с MCP сервером,
 * гарантируя 100% совместимость с FastMCP серверами.
 *
 * Требования:
 * - Python 3.7+
 * - fastmcp библиотека установлена: pip install fastmcp
 * - mcp_python_client.py в корне проекта
 */
class MCPPythonClient(
    private val serverUrl: String
) : MCPClientInterface, AutoCloseable {

    private val pythonScript = File("mcp_python_client.py").absolutePath
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    /**
     * Проверяет наличие Python и необходимых зависимостей
     */
    private fun checkPythonAvailability(): Result<String> {
        return try {
            // Пробуем разные команды Python
            val pythonCommands = listOf("python3", "python")

            for (cmd in pythonCommands) {
                try {
                    val process = ProcessBuilder(cmd, "--version")
                        .redirectErrorStream(true)
                        .start()

                    val exitCode = process.waitFor()
                    if (exitCode == 0) {
                        return Result.success(cmd)
                    }
                } catch (e: Exception) {
                    // Пробуем следующую команду
                }
            }

            Result.failure(Exception("Python не найден. Установите Python 3.7+"))
        } catch (e: Exception) {
            Result.failure(Exception("Ошибка проверки Python: ${e.message}"))
        }
    }

    /**
     * Выполняет Python скрипт с заданными аргументами
     */
    private suspend fun executePythonScript(vararg args: String): Result<JsonObject> = withContext(Dispatchers.IO) {
        try {
            // Проверяем наличие Python
            val pythonCmd = checkPythonAvailability().getOrElse {
                return@withContext Result.failure(it)
            }

            // Проверяем наличие скрипта
            if (!File(pythonScript).exists()) {
                return@withContext Result.failure(
                    Exception("Python скрипт не найден: $pythonScript")
                )
            }

            println("🐍 Запуск Python клиента: $pythonCmd $pythonScript ${args.joinToString(" ")}")

            // Запускаем Python скрипт
            val processBuilder = ProcessBuilder(pythonCmd, pythonScript, *args)
                .redirectErrorStream(false)

            val process = processBuilder.start()

            // Читаем stdout
            val output = BufferedReader(InputStreamReader(process.inputStream)).use { reader ->
                reader.readText()
            }

            // Читаем stderr
            val error = BufferedReader(InputStreamReader(process.errorStream)).use { reader ->
                reader.readText()
            }

            val exitCode = process.waitFor()

            println("📊 Python вывод: $output")
            if (error.isNotBlank()) {
                println("⚠️ Python stderr: $error")

                // Проверяем специфичные ошибки
                if (error.contains("No module named 'fastmcp'")) {
                    println("💡 Подсказка: Установите fastmcp: pip install fastmcp")
                    println("💡 Или позвольте скрипту установить его автоматически")
                }
            }

            if (exitCode != 0) {
                val errorMsg = if (error.contains("No module named 'fastmcp'")) {
                    "fastmcp не установлен. Установите командой: pip install fastmcp\n\nПолная ошибка: $error"
                } else {
                    "Python скрипт завершился с ошибкой (код $exitCode): $error"
                }
                return@withContext Result.failure(Exception(errorMsg))
            }

            // Парсим JSON результат
            try {
                val result = json.decodeFromString<JsonObject>(output.trim())
                Result.success(result)
            } catch (e: Exception) {
                Result.failure(
                    Exception("Не удалось распарсить ответ Python: $output. Ошибка: ${e.message}")
                )
            }

        } catch (e: Exception) {
            Result.failure(Exception("Ошибка выполнения Python скрипта: ${e.message}", e))
        }
    }

    override suspend fun initialize(): Result<MCPInitializeResult> = withContext(Dispatchers.IO) {
        try {
            val result = executePythonScript("initialize", serverUrl).getOrElse {
                return@withContext Result.failure(it)
            }

            val success = result["success"]?.jsonPrimitive?.boolean ?: false
            if (!success) {
                val error = result["error"]?.jsonPrimitive?.content ?: "Неизвестная ошибка"
                return@withContext Result.failure(Exception("Python initialize error: $error"))
            }

            // Создаем MCPInitializeResult из ответа
            val initResult = MCPInitializeResult(
                protocolVersion = result["protocolVersion"]?.jsonPrimitive?.content ?: "2024-11-05",
                serverInfo = MCPServerInfo(
                    name = result["serverInfo"]?.jsonObject?.get("name")?.jsonPrimitive?.content ?: "MCP Server",
                    version = result["serverInfo"]?.jsonObject?.get("version")?.jsonPrimitive?.content ?: "1.0.0"
                ),
                capabilities = JsonObject(emptyMap())
            )

            Result.success(initResult)
        } catch (e: Exception) {
            Result.failure(Exception("Ошибка инициализации через Python: ${e.message}", e))
        }
    }

    override suspend fun listTools(): Result<List<MCPTool>> = withContext(Dispatchers.IO) {
        try {
            // Для получения списка инструментов вызываем ttools
            val ttoolsResult = callTool("ttools", emptyMap()).getOrElse {
                // Если ttools не доступен, возвращаем пустой список
                return@withContext Result.success(emptyList())
            }

            // Парсим результат ttools
            val ttoolsText = ttoolsResult.content.firstOrNull()?.text ?: "[]"

            try {
                val toolsArray = json.decodeFromString<JsonArray>(ttoolsText)
                val tools = toolsArray.mapNotNull { element ->
                    try {
                        val obj = element.jsonObject
                        val name = obj["name"]?.jsonPrimitive?.content ?: return@mapNotNull null
                        val signature = obj["signature"]?.jsonPrimitive?.content ?: ""
                        val doc = obj["doc"]?.jsonPrimitive?.content ?: ""

                        MCPTool(
                            name = name,
                            description = "$signature${if (doc.isNotEmpty()) " - $doc" else ""}",
                            inputSchema = null
                        )
                    } catch (e: Exception) {
                        null
                    }
                }

                Result.success(tools)
            } catch (e: Exception) {
                println("⚠️ Не удалось распарсить ttools: ${e.message}")
                Result.success(emptyList())
            }
        } catch (e: Exception) {
            Result.failure(Exception("Ошибка получения списка инструментов: ${e.message}", e))
        }
    }

    override suspend fun callTool(toolName: String, arguments: Map<String, Any>): Result<MCPToolCallResult> = withContext(Dispatchers.IO) {
        try {
            // Простое преобразование Map в JSON строку
            val argsJson = buildJsonObject {
                arguments.forEach { (key, value) ->
                    when (value) {
                        is String -> put(key, value)
                        is Number -> put(key, value)
                        is Boolean -> put(key, value)
                        else -> put(key, value.toString())
                    }
                }
            }.toString()

            val result = executePythonScript("call_tool", serverUrl, toolName, argsJson).getOrElse {
                return@withContext Result.failure(it)
            }

            val success = result["success"]?.jsonPrimitive?.boolean ?: false
            if (!success) {
                val error = result["error"]?.jsonPrimitive?.content ?: "Неизвестная ошибка"
                return@withContext Result.failure(Exception("Python call_tool error: $error"))
            }

            // Преобразуем результат в MCPToolCallResult
            val content = result["content"]?.jsonArray?.mapNotNull { element ->
                try {
                    val obj = element.jsonObject
                    MCPContent(
                        type = obj["type"]?.jsonPrimitive?.content ?: "text",
                        text = obj["text"]?.jsonPrimitive?.content,
                        data = obj["data"]?.jsonPrimitive?.content,
                        mimeType = obj["mimeType"]?.jsonPrimitive?.content
                    )
                } catch (e: Exception) {
                    null
                }
            } ?: emptyList()

            Result.success(MCPToolCallResult(content = content, isError = false))
        } catch (e: Exception) {
            Result.failure(Exception("Ошибка вызова инструмента '$toolName': ${e.message}", e))
        }
    }

    override fun close() {
        // Python клиент не требует закрытия
    }
}

File: mcp_python_client.py
#!/usr/bin/env python3
"""
Python MCP клиент для использования из Kotlin приложения
Использует FastMCP для гарантированной совместимости с локальным сервером
"""

import asyncio
import json
import sys
import subprocess

# Проверяем и устанавливаем fastmcp при необходимости
try:
    from fastmcp import Client
except ImportError:
    print(json.dumps({
        "success": False,
        "error": "fastmcp не установлен. Попытка автоматической установки...",
        "action": "installing"
    }), file=sys.stderr)

    try:
        # Пытаемся установить fastmcp
        subprocess.check_call([sys.executable, "-m", "pip", "install", "fastmcp", "--quiet"])
        print(json.dumps({
            "success": True,
            "message": "fastmcp успешно установлен"
        }), file=sys.stderr)
        from fastmcp import Client
    except Exception as e:
        print(json.dumps({
            "success": False,
            "error": f"Не удалось установить fastmcp: {str(e)}. Выполните: pip install fastmcp"
        }))
        sys.exit(1)

async def initialize(server_url: str) -> dict:
    """Инициализация соединения с MCP сервером"""
    client = Client(server_url)
    async with client:
        # Получаем информацию о сервере через initialize
        # FastMCP делает это автоматически при входе в контекст
        return {
            "success": True,
            "serverInfo": {
                "name": "MCP Server",
                "version": "1.0.0"
            },
            "protocolVersion": "2024-11-05"
        }

async def list_tools(server_url: str) -> dict:
    """Получение списка инструментов"""
    client = Client(server_url)
    try:
        async with client:
            # FastMCP автоматически получает список инструментов
            tools = []
            # Пытаемся получить информацию через inspect
            # или просто возвращаем пустой список
            return {
                "success": True,
                "tools": tools
            }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }

async def call_tool(server_url: str, tool_name: str, arguments: dict) -> dict:
    """Вызов инструмента MCP сервера"""
    client = Client(server_url)
    try:
        async with client:
            result = await client.call_tool(tool_name, arguments)

            # Извлекаем правильное значение из CallToolResult
            # result.content - это список TextContent объектов
            content_list = []
            if hasattr(result, 'content') and result.content:
                for item in result.content:
                    if hasattr(item, 'text') and item.text:
                        content_list.append({
                            "type": "text",
                            "text": item.text
                        })

            # Если ничего не нашли, используем str(result) как fallback
            if not content_list:
                content_list = [{
                    "type": "text",
                    "text": str(result)
                }]

            return {
                "success": True,
                "content": content_list
            }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }

def main():
    """Главная функция для обработки команд"""
    if len(sys.argv) < 3:
        print(json.dumps({
            "success": False,
            "error": "Usage: mcp_python_client.py <command> <server_url> [args...]"
        }))
        sys.exit(1)

    command = sys.argv[1]
    server_url = sys.argv[2]

    try:
        if command == "initialize":
            result = asyncio.run(initialize(server_url))
            print(json.dumps(result))

        elif command == "list_tools":
            result = asyncio.run(list_tools(server_url))
            print(json.dumps(result))

        elif command == "call_tool":
            if len(sys.argv) < 4:
                print(json.dumps({
                    "success": False,
                    "error": "Tool name required"
                }))
                sys.exit(1)

            tool_name = sys.argv[3]
            arguments = json.loads(sys.argv[4]) if len(sys.argv) > 4 else {}

            result = asyncio.run(call_tool(server_url, tool_name, arguments))
            print(json.dumps(result))

        else:
            print(json.dumps({
                "success": False,
                "error": f"Unknown command: {command}"
            }))
            sys.exit(1)

    except Exception as e:
        print(json.dumps({
            "success": False,
            "error": str(e)
        }))
        sys.exit(1)

if __name__ == "__main__":
    main()
